src/mlpack/methods/ann/layer/gru_impl.hpp

```diff
@@ -0,0 +1,307 @@
+
+  // Delta rt.
+  arma::mat d_rt = boost::apply_visitor(deltaVisitor, outputHidden2GateModule) %
+      *backIterator;
+
+  // Delta of forget gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, forgetGateModule)), std::move(d_rt),
+      std::move(boost::apply_visitor(deltaVisitor, forgetGateModule))),
+      forgetGateModule);
+
+  // Put delta zt.
+  prevError.submat(0, 0, 1 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, inputGateModule);
+
+  // Put delta rt.
+  prevError.submat(1 * outSize, 0, 2 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, forgetGateModule);
+
+  // Put delta ot.
+  prevError.submat(2 * outSize, 0, 3 * outSize - 1, 0) = boost::apply_visitor(
+      deltaVisitor, hiddenStateModule);
+
+  // Get delta ht - 1 for input gate and forget gate.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, input2GateModule)), 
+      std::move(prevError.submat(0, 0, 2 * outSize - 1, 0)),
+      std::move(boost::apply_visitor(deltaVisitor, output2GateModule))),
+      output2GateModule);   
+
+  // Add delta ht - 1 from hidden state.
+  boost::apply_visitor(deltaVisitor, output2GateModule) += 
+      boost::apply_visitor(deltaVisitor, outputHidden2GateModule) %
+      boost::apply_visitor(outputParameterVisitor, forgetGateModule);
+
+  // Add delta ht - 1 from ht.
+  boost::apply_visitor(deltaVisitor, output2GateModule) += gy %
+      boost::apply_visitor(outputParameterVisitor, inputGateModule);
+
+  // Get delta input.
+  boost::apply_visitor(BackwardVisitor(std::move(boost::apply_visitor(
+      outputParameterVisitor, input2GateModule)), std::move(prevError),
+      std::move(boost::apply_visitor(deltaVisitor, input2GateModule))),
+      input2GateModule);
+
+  backwardStep++;
+  backIterator--;
+
+  g = boost::apply_visitor(deltaVisitor, input2GateModule);
+}
+
+template<typename InputDataType, typename OutputDataType>
+template<typename eT>
+void GRU<InputDataType, OutputDataType>::Gradient(
+    arma::Mat<eT>&& input,
+    arma::Mat<eT>&& /* error */,
+    arma::Mat<eT>&& /* gradient */)
+{
+  if (gradIterator == outParameter.end())
+  {
+    gradIterator = --(--outParameter.end());
+  }
+
+  boost::apply_visitor(GradientVisitor(std::move(input), std::move(prevError)),
+      input2GateModule);
+
+  boost::apply_visitor(GradientVisitor(
+      std::move(*gradIterator),
+      std::move(prevError.submat(0, 0, 2 * outSize - 1, 0))), 
+      output2GateModule);
+
+  boost::apply_visitor(GradientVisitor(
+      std::move(*gradIterator),
+      std::move(prevError.submat(2 * outSize, 0, 3 * outSize - 1, 0))), 
+      outputHidden2GateModule);
+
+  gradIterator--;
+}
+
+template<typename InputDataType, typename OutputDataType>
+void GRU<InputDataType, OutputDataType>::ResetCell()
+{
+  outParameter.clear();
+  outParameter.push_back(arma::zeros<arma::mat>(outSize, 1));
```

Comment:
Correct me if I am misunderstanding---I think the first element of `outParameter` will always be a matrix of zeros and it is never modified.  This leads me to three thoughts:

1) If `outParameter.front()` is always `arma::zeros<arma::mat>(outSize, 1)` and `outSize` does not change between calls to `ResetCell()` we can just do:

```
arma::mat tmp = std::move(outParameter[0]);
outParameter.clear();
outParameter.push_back(arma::mat()); // not 100% sure on the syntax here, I think this is right
outParameter.front() = std::move(tmp); // maybe this could be compressed with the previous line but you get the idea
```

2) In another place you also push back a zeros matrix onto `outParameter`.  If `outParameter` held pointers to matrices instead, you could do `outParameter.push_back(outParameter.front())`.  But you could also create an alias matrix with one of the special Armadillo constructor, of `outParameter.front()`, and push that onto the list.

3) Would it be possible to avoid allocating the zeros matrix entirely?  And then detect when you are at the front of `outParameter`?  You use `outParameter` in code like...

```
arma::mat d_zt = gy % (*backIterator - stuff);
```

but if we could instead do something like

```
if (we are at the first element)
  arma::mat d_zt = gy % stuff;
else
  arma::mat d_zt = gy % (*backIterator - stuff);
```

then we can save a possibly significant amount of computation.  Since we are pushing the zeros matrix into something of type `arma::mat`, it is no longer known at compile time that the matrix only has zeros.  Thus Armadillo is not able to see something like `*backIterator - stuff` and know that `*backIterator` is actually a matrix of zeros, and an unnecessary loop over the values happens.

The third comment is the most complex of the three options, but it will result in the most speedup so I think it would be worth it to implement.  If you don't want to do that now, we should open an issue for it.  But at the very least we should do #1 (and #2) if not, plus possibly the improvement suggested by Marcus above.